benches/search_comparison.rs

```diff
@@ -285,3 +255,50 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
 fn search_sorted_vec<'a, T: Ord>(c: &'a Vec<T>, x: T) -> Option<&'a T> {
     c.binary_search(&x).ok().map(|i| &c[i])
 }
+
+/// Generate pseudorandom sequence of numbers
+///
+/// ```rust
+/// assert_eq!(
+///     pseudorandom_iter::<u32>(0, u32::MAX , None).take(16).collect(),
+///     vec![2027808446, 2393657406, 900912232, 833811770, 1061328792, 93432844,
+///     2565420364, 1550801266, 1147887774, 710446582, 3306204668, 500014398,
+///     1140212266, 2163551532, 513205252, 1774545590]
+///
+/// )
+/// ```
+///
+/// ```rust
+/// assert_eq!(
+///     pseudorandom_iter::<u8>(0, u8::MAX, None).take(16).collect(),
+///     vec![250, 200, 36, 20, 178, 78, 4, 152, 174, 8, 128, 198, 166, 206, 156, 80]
+/// );
+/// ```
+///
+/// ```rust
+/// assert_eq!(
+///     pseudorandom_iter::<u8>(0, u8::MAX, Some(32)).take(16).collect(),
+///     vec![27, 9, 5, 20, 18, 15, 5, 24, 15, 8, 1, 6, 7, 15, 29, 17]
+/// );
+fn pseudorandom_iter<T>(
+    mut seed: usize,
+    max: usize,
+    bound: Option<usize>,
+) -> impl Iterator<Item = T>
+where
+    T: TryFrom<usize>,
+    <T as TryFrom<usize>>::Error: core::fmt::Debug,
+{
+    std::iter::from_fn(move || {
+        // LCG constants from https://en.wikipedia.org/wiki/Numerical_Recipes.
+        seed = seed.wrapping_mul(1664525).wrapping_add(1013904223);
+        let r = (seed.wrapping_mul(2)) % max;
+        let r = if let Some(bound) = bound {
+            r % bound
+        } else {
+            r - r % 2
```

Comment:
Doesn't this mean we could have avoided the `*2` when going into this branch?

Reply:
Maybe we can eliminate the whole this if-else altogether? "Bound"-computations are covered by `max` parameter of the function and duplication of values is now handled by the clients of this function. So AFAIU `pseudorandom_iter()` does not need to worry about duplicates anymore.

Reply:
`bound` and `max` are different. 

In search case, for example, for `u32 size=512` we will have `MAX=2^32`, 
- search array `v` is generated from iterator `(0..).map(|i| i * 2)` with will have `1024` as a highest value
- but `pseudorandom_iter` will keep generating you values in the whole 0..2^32 range 
     > [1013904223, 1196828703, 450456268, 2817872298, 2081980138, 2971988253, 2135380753, 987533808]


So, how would you prefer to handle it? 
1. keep `bound` as a input parameter in the `pseudorandom_iter` 
2. do `let x = r.next().unwrap() % T::try_from((size * 2) % max).unwrap();` with `T: Rem<Output=T> `

Reply:
removed `if-else` entirely